### PR TITLE
[Task-07] Max Snischuk - Api requests

### PR DESCRIPTION
**Pull request by snischuk**
***[Link to deployment](https://snischuk.github.io/fe-react-2024/)***

***Acceptance criteria for 1***

- [x] Mentor is added to pr
- [x] AC copied to pr description
- [x] Link to deployed version is added to pr
- [x] `ProductPage` fetches data from the API using either `fetch` or `axios`
- [x] Pagination is implemented using query parameters
- [x] Search functionality is implemented using query parameters, with the search triggered by a search button
- [x] Category filtering is implemented using query parameters, resetting the page to the first page on change
- [x] A loading spinner or message is displayed while awaiting API requests
- [x] If the API response is an empty array or an error occurs, a "No products" message is displayed instead of a blank screen
 
***Acceptance criteria for 1****

- [x] Pagination is hidden on mobile view
- [x] Infinite scroll is implemented for mobile view, either manually or using a library
- [x] Error notifications are implemented using axios interceptors, with component libraries for toastr notifications

### CODE REVIEW

File: src/components/ControlPanel/ControlPanel.tsx [hunk 1]
```diff
@@ -1,33 +1,41 @@
...
 
-    const getFilterButtonClassName = (filterButtonName: ProductCategoryName) =>
-        selectedFiltersByCategory.includes(filterButtonName) ? `${styles.filterBtn} ${styles.filterBtnActive}` : styles.filterBtn;
+    const onCategoryClick: MouseEventHandler<HTMLButtonElement> = (event) => {
+        const clickedFilterOptionId = (event.target as HTMLButtonElement).dataset.categoryId;
+        const filterOption = filtersOptionsByCategory.find((option) => option.id.toString() === clickedFilterOptionId);
+        if (filterOption) {
+            onFilterByCategoryClick(filterOption as ProductFilterByCategory);
+        }
+    };
```
Comment (vladyslav-yermolin-moc) [inline src/components/ControlPanel/ControlPanel.tsx:42]: The text in the search pannel is disappear on search submit
image

Comment (vladyslav-yermolin-moc) [reply src/components/ControlPanel/ControlPanel.tsx:42]: Also would be good to add searching by Enter key in this field

**Review by vladyslav-yermolin-moc — CHANGES_REQUESTED** [review]
Hey, @snischuk, sorry for a long review. Good work 👍

**Comment by snischuk** [comment]
I saw in Myron's `services` folder all work with API requests is done in OOP style. Is it a best practice to write services in OOP style (it confused me because my whole project is written in functional style)?

File: src/components/ControlPanel/ControlPanel.tsx (hunk 1 above)
Comment (snischuk) [reply src/components/ControlPanel/ControlPanel.tsx:42]: fixed

**Review by vladyslav-yermolin-moc — APPROVED** [review]

**Comment by vladyslav-yermolin-moc** [comment]
Hey, @snischuk, actually OOP in the services is kind of an old approach. At the moment, in the React world common approach is a functional paradigm of programming. 
1. You create an Axios instance, set for it default params, baseUrl, headers, etc, and export this instance. https://axios-http.com/docs/config_defaults. Also you can add some middleware for errors or authorization.
2. The next layer is your services, which use the Axios instance with needed params. It could be a simple function that you export:
```ts
export const getCategories = () => axiosInstance.get('categories')
```

3. And in your component you utilize these services.

I think, OOP services, which you implemented here were a good practice for you.

P.s: I'm on vacation, so could miss the notification of your homework

**Pull request merged**